### PR TITLE
#211 Credentials extraction

### LINKED ISSUE #211: Authorization: implement extraction of user credentials from the request
Extract token or user/password from gRPC and REST requests and reject/accept that request based on on response from credential validator

### PR DESCRIPTION
**Pull request by vovac12**
Closes #211

### CODE REVIEW

File: bob-access/src/credentials.rs [hunk 1]
```diff
@@ -1,2 +1,41 @@
...
+        password: impl Into<String>,
+    ) -> Self {
+        self.kind = Some(CredentialsKind::Basic {
+            username: username.into(),
+            password: password.into(),
+        });
+        self
+    }
+
+    pub fn with_token(mut self, token: impl Into<String>) -> Self {
```
Comment (idruzhitskiy) [inline bob-access/src/credentials.rs:32]: IMO there should be either separate builder for the credentials, or multiple constructors

File: bob-access/src/extractor.rs [hunk 2]
```diff
@@ -14,7 +16,112 @@ impl StubExtractor {
...
+pub struct BasicExtractor {}
+
+impl<T> Extractor<TRequest<T>> for BasicExtractor {
+    fn extract(&self, req: &TRequest<T>) -> Result<Credentials, Error> {
+        let meta = req.metadata();
+        let username = meta
+            .get("username")
+            .ok_or_else(|| Error::credentials_not_provided("username"))?
+            .to_str()
+            .map_err(|e| Error::unknown(e.to_string()))?;
```
Comment (idruzhitskiy) [inline bob-access/src/extractor.rs:34]: Why unknown? Probably should be "parse error" or "conversion error"

File: bob-access/src/extractor.rs [hunk 3]
```diff
@@ -14,7 +16,112 @@ impl StubExtractor {
...
+pub struct MultiExtractor {
+    basic_extractor: BasicExtractor,
+    token_extractor: TokenExtractor,
+}
+
+impl<T> Extractor<TRequest<T>> for MultiExtractor {
+    fn extract(&self, req: &TRequest<T>) -> Result<Credentials, Error> {
+        let basic_credentials = self.basic_extractor.extract(req);
+        let token_credentials = self.token_extractor.extract(req);
+        if basic_credentials.is_ok() && token_credentials.is_ok() {
```
Comment (idruzhitskiy) [inline bob-access/src/extractor.rs:101]: IMO pattern matching with `match (basic_credentials.is_ok(), token_credentials.is_ok()) { (true, true) => ... }` is more readable

File: bob-access/src/credentials.rs (hunk 1 above)
Comment (vovac12) [reply bob-access/src/credentials.rs]: I added a `CredentialsBuilder`

File: bob-access/src/extractor.rs (hunk 2 above)
Comment (vovac12) [reply bob-access/src/extractor.rs]: Replaced with `ConversionError`

File: bob-access/src/extractor.rs (hunk 3 above)
Comment (vovac12) [reply bob-access/src/extractor.rs:105]: Done

Comment (pyakushin) [reply bob-access/src/extractor.rs:105]: @vovac12 You added this pattern only for `MultiExtractor`

**Review by pyakushin — APPROVED** [review]
LGTM, check 1 comment

**Comment by vovac12** [comment]
@pyakushin Fixed

**Comment by idruzhitskiy** [comment]
The "famous" match part in extractor impls can be factored out in smth like "find_single_credentials", but that's not necessary. Other than that - LGTM.

**Pull request merged**